src/main/java/me/roinujnosde/titansbattle/BaseGame.java

```diff
@@ -0,0 +1,707 @@
+                reason = plugin.getLang("maximum-groups", this);
+                plugin.debug("max groups");
+            }
+            Integer amountOfPlayers = getGroupParticipants().getOrDefault(warrior.getGroup(), 0);
+            if (amountOfPlayers >= getConfig().getMaximumPlayersPerGroup() &&
+                    getConfig().getMaximumPlayersPerGroup() > 0) {
+                reason = plugin.getLang("maximum-players-per-group", this);
+                plugin.debug("max per group");
+            }
+        }
+        if (getConfig().isUseKits() && Kit.inventoryHasItems(player)) {
+            reason = plugin.getLang("clear-your-inventory", this);
+            plugin.debug("clear inv");
+        }
+
+        PlayerJoinGameEvent event = new PlayerJoinGameEvent(player, this);
+        Bukkit.getPluginManager().callEvent(event);
+        plugin.debug("cancel: " + event.isCancelled());
+        plugin.debug("reason: >>>" + reason + "<<<");
+        if (reason != null) {
+            player.sendMessage(reason);
+        }
+        return reason == null && !event.isCancelled();
+    }
+
+    protected void processPlayerExit(@NotNull Warrior warrior) {
+        if (!isParticipant(warrior)) {
+            return;
+        }
+        Player player = warrior.toOnlinePlayer();
+        if (player != null) {
+            teleport(warrior, getConfig().getExit());
+            PlayerExitGameEvent event = new PlayerExitGameEvent(player, this);
+            Bukkit.getPluginManager().callEvent(event);
+        }
+        participants.remove(warrior);
+        Group group = warrior.getGroup();
+        if (!isLobby()) {
+            runCommandsAfterBattle(Collections.singletonList(warrior));
+            processRemainingPlayers(warrior);
+            //last participant
+            if (getConfig().isGroupMode() && group != null && !getGroupParticipants().containsKey(group)) {
+                broadcastKey("group_defeated", group.getName());
+                Bukkit.getPluginManager().callEvent(new GroupDefeatedEvent(group, warrior.toOnlinePlayer()));
+                group.getData().increaseDefeats(getConfig().getName());
+            }
+            sendRemainingOpponentsCount();
+        }
+    }
+
+    protected abstract void processRemainingPlayers(@NotNull Warrior warrior);
+
+    protected void setKit(@NotNull Warrior warrior) {
+        Player player = warrior.toOnlinePlayer();
+        Kit kit = getConfig().getKit();
+        if (getConfig().isUseKits() && kit != null && player != null) {
+            Kit.clearInventory(player);
+            kit.set(player);
+        }
+    }
+
+    protected void playDeathSound(@NotNull Warrior victim) {
+        Stream<Player> players = getPlayerParticipantsStream();
+        if (!getConfig().isGroupMode()) {
+            players.forEach(p -> SoundUtils.playSound(ENEMY_DEATH, plugin.getConfig(), p));
+            return;
+        }
+        GroupManager groupManager = plugin.getGroupManager();
```

Comment:
Group**s**Manager? 👀

Reply:
ClansManager?

Also can't change it now: API.